[integrations][Cloudflare Logpull] - Fixed invalid time range issue

ShourieG opened the pull request:
## Type of change
- Bug

## What does this PR do?
This PR uses a certain time range constraint discussed [here](https://github.com/elastic/integrations/issues/4414) to address the issue. The main concept lies in the following time constraints : 
```js
now - start <= 168h
now - end > 1m
start < end
end - stard <= 1h
 
start = max(time_of_last_event, now - 168h)  # This Always satisfies the 168h requirement.
end = min(start + 1h, now - 1m)  # Always satisfies the 1h window requirement, and never requests earlier data.
```
The integration uses the above constraints while calculating the value of the **start & end** query params. 

## NOTE: 
1) HTTPJSON does not have any mechanism to store values internally while setting parameters via the config. It also does not have the capability to pull values from the **.url.***  objects, only .**last_response.url.*** works. This creates a problem we need to work around. For the calculation of **end** using the logic **end = min(start + 1h, now - 1m)**, the value of **start** needs to be calculated here since we cannot use **[[.url.params.Get "start"]]**. Similarly we are unable to store a singular value for **now** and it gets computed everywhere it's used. This creates very minor deviations in the timestamps of a few milliseconds to maybe a max of 1 second. We can probably live by this. This re-calculation of values effectively transforms the internal logic for computing the value of **end** to:
```js
end = min(max(time_of_last_event, now - 168h) + 1h, now - 1m)
```
2) This was tested manually with a local server and all of the constraints passed. 
3) A mock service has been integrated via docker-compose which simulates the exact api constraints during the system test 
     with a hit_count of 3 responses. The service sends a rolling response similar to elastic stream http servers.

## Checklist
- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues
- Closes #4414



## Screenshots
<img width="580" alt="cf_logbull_tests" src="https://github.com/elastic/integrations/assets/105607378/0dabf5d2-51d3-41bd-acc2-06350d82b9f2">

elasticmachine commented:
Pinging @elastic/security-external-integrations (Team:Security-External Integrations)

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-7726/detail/PR-7726/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-7726/detail/PR-7726/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-7726/detail/PR-7726/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-7726/detail/PR-7726/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_7726.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-09-08T13:09:48.208Z&rangeTo=2023-09-08T13:29:48.208Z&transactionName=BUILD+Ingest-manager%2Fintegrations%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=1f6b4a9ca4a4b405eae783287f944d3e&transactionId=a42f4fe40e4a9a1e)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-09-08T13:19:48.208+0000


* Duration: 24 min 14 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![3](https://img.shields.io/badge/3%20-red)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-7726/detail/PR-7726/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Test integration: cloudflare`
<ul>
<li>Took 11 min 2 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-7726/runs/1/steps/1201/log/?start=0">here</a></li>
<li>Description: <code>                                eval "$(../../build/elastic-package stack shellinit)"                                ../../build/elastic-package test -v --report-format xUnit --report-output file --test-coverage                                </code></l1>
</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-7726/runs/1/steps/1290/log/?start=0">here</a></li>

</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-7726/runs/1/steps/1303/log/?start=0">here</a></li>

</ul>
  
  </p>
  </details>



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 100.0% (`2/2`) | :green_heart: 
Files | 100.0% (`3/3`) | :green_heart: 
Classes | 100.0% (`3/3`) | :green_heart: 
Methods | 100.0% (`36/36`) | :green_heart: 24.242
Lines | 94.73% (`701/740`) | :-1: -5.27
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

efd6 reviewed (commented):
You'll need to bump the stack version to pick up the `max` and `min` functions.

ShourieG commented:
@andrewkroh I've added a mock service that simulates the api constraints during system tests and resolved rest of the suggestions.

andrewkroh commented on packages/cloudflare/_dev/deploy/docker/logpull-mock-service/main.go:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	// we have a valid timestamp range, now return a rolling response
+	responses := []string{
```
This can be pulled out of the method and stored as a global.

andrewkroh commented on packages/cloudflare/_dev/deploy/docker/logpull-mock-service/main.go:
```diff
@@ -0,0 +1,129 @@
+	// apply constraints on the timestamp values
+	now := time.Now().UTC()
+	if now.Sub(startTimestamp) > 168*time.Hour {
+		http.Error(w, `{"message":"constraint failed: now - start <= 168h"}`, http.StatusBadRequest)
```
Logging is not consistently applied on errors (the previous ones were logged, this isn't). I suggest a helper to consistently log the error and write the error. Like perhaps...

```
func httpError(w http.ResponseWriter, r *http.Request, error string, code int) {
	log.Printf("%s - WARN Returning %d: %s", r.RemoteAddr, code, error)
	http.Error(w, error, code)
}
```

andrewkroh reviewed (commented):
The mock server you added is helpful, but it's hard to prove the httpjson config is working correctly because it does not appear that the min/max range protections get exercised. So I think it would be good if we can find a way to test these cases.

ShourieG commented on packages/cloudflare/_dev/deploy/docker/logpull-mock-service/main.go:
```diff
@@ -0,0 +1,129 @@
+
+	port := 3000
+	log.Printf("Server listening on port %d...\n", port)
+	router.Use(mux.CORSMethodMiddleware(router))
```
yea wil remove this

ShourieG commented on packages/cloudflare/data_stream/logpull/_dev/test/system/test-default-config.yml:
```diff
@@ -1,12 +1,15 @@
   enable_request_tracer: true
 data_stream:
   vars:
+    interval: 1h
```
This was caused due to a template error, fixed this in the latest update.

ShourieG commented:
@andrewkroh resolved all the suggested changes, introduced a new test that works over 2 iterations so cursor value is used. Modified the server code to match a specific criteria for the new test case to validate cursor value. I'm not sure how we can add more unique cursor tests without having a way to dynamically inject a httpjson.yml.hbs or somehow modify the agent yml code on the fly.

elasticmachine commented:
Package cloudflare - 2.20.0 containing this change is available at https://epr.elastic.co/search?package=cloudflare
